### ISSUE TITLE
PROPOSAL: Peer review

### ISSUE THREAD

**Issue by talvo** [pennmush/pennmush#1229]
As the dev team has gotten both larger (hi Qon!) and more active recently, I think it's more important than ever that we have active discussion about the direction Penn goes in, and agreement on how best to get there.

I would therefore like to propose that any code change which isn't a bug fix be added as a Pull Request, to be reviewed (or, at the very least, glanced at) by another member of the dev team who can then push it to master. This will hopefully provoke more discussion, and give a better chance for changes to be reviewed and collaborated on. It will also allow us to share ideas for/agree on the best approach to things before a new feature is merged, rather than it being released and then immediately tinkered with as someone else has additional thoughts/ideas for features/syntax, etc, but missed the commit and didn't get to chime in before it went out.

What do you guys think? @captdeaf, @shawn, @mike347.

**Comment by HarryCordewener** [comment]
I am not part of the dev-team, but that should be the standard for any Pull Requests. At least one approval on the PR.

**Comment by talvo** [comment]
The point is that most things are not done as PRs in the first place, they're simply pushed to master.

**Comment by mike347** [comment]
I agree.. if only one person is hacking away at things, then the.. 'old method?' of just pushing to master works. Raev has done a good job of keeping things in a separate branch while testing, rather than pushing to master, but I definitely think having to do a PR .. and being 'seconded' by another Dev is a good idea. 

The 'Penn as a webserver' patch, which we now have two of.. one from Walker, and one from grapenut, is a good example of how having a PR and discussing it before pulling it could be beneficial.. We've got two implementations of the same idea, both with their own pros and cons. That is a -good- thing, because we can pick the good from both and get rid of the cons. That only happens with discussion and thought.

**Comment by ray73864** [comment]
Just my 2 cents, but, you can set the master branch to 'protected'. With the rules you can then also make it so it requires multiple reviews, etc...

https://help.github.com/articles/configuring-protected-branches/

**Comment by captdeaf** [comment]
Agreed - I've tried making all my non-bugfix changes into PRs. HTTPd, `@trigger/inline`, Millisecond timing, one-queue, etc. Small bugfix patches I'm less concerned about, though.

For any kind of new feature, or anything not explicitly a bugfix, yeah, I vote a PR with at least one other dev to approve it. Doesn't necessarily have to be a full on code review (though that would be a good thing), as waiting for that can hang development, but "Sounds good, push away" would be good.

Then again, I may have just shot myself in the foot preventing me from sneaking in a commit to remove lit() ... :D.

**Comment by MrWigggles** [comment]
This is seems like a smart move.

**Comment by tkrajcar** [comment]
I'm a big fan of this idea, but would like to remove the 'this applies to things that aren't a bugfix' clause and apply it to all changes. All changes to master should be sidekicked.

We should use the PR approval feature built into Github, require a single approver, and protect the master branch. That way people can still collaborate together on other branches, and there's no worry about deciding whether a change is 'exempt from the rules' or not.

If we begin having issues with stale PRs, I'm open to revisiting this, but I also think we all communicate enough that a well-placed @ or PR assignment in Github or a shout on `M*U*S*H` will do the job.

**Comment by grapenut** [comment]
Routing all changes through the PR system would increase the visibility for contributed PRs from non-devs like myself as well. More eyes on the PR queue should keep them from going stale. I support 100%.

**Comment by captdeaf** [comment]
I was thinking how I don't want to wait for PRs for bugfixes for M*U*S*H (e.g: The infinite dbsaves bug in millisecond timers) when I realized that was a bit backwards.

Many contributors have live mushes where they test their changes. I think Shawn, Mike G and I tend to develop on barebones mushes, then push to M*U*S*H - it's supposedly our testbed. But unlike contributors, we push to `master` to get it to our "live testbed". That's a bad idea.

M*U*S*H claims to be a testbed - we need to make it an actual testbed to exist and be used _before_ `master` branch.

Our current workflow for getting code onto M*U*S*H:

1. push changes to `master`
2. Log in to server, sudo, `git fetch`, merge master. (There's a few HC changes in M*U*S*H's local checkout)
3. Compile, `> @shutdown/reboot`

We should probably change that to, say, a `development` branch? And make PRs for any changes we want to commit to master? I'd also say `development` branch shouldn't be expected to retain history in that case as we'd have to frequently change it to a clean `master` checkout when no PRs are being tested.

We could go with something more formal, sure, but we're an informal, tiny group, not a startup agile team X_X.

**Comment by HarryCordewener** [comment]
I am all for the development branch!

**Comment by captdeaf** [comment]
Assigning to @shawnw for comment

**Comment by shawnw** [comment]
I'm okay with PRs for big stuff - anything you'd do in a branch anyways, that involves multiple commits, or things you want feedback on first (Hello pcre2 now being a much smaller patch thanks to that). Bug fixes and trivial changes? Meh.

I don't know about a specific development branch instead of master... I like the idea of a stable branch like 1.8.7-bugfixes is instead, but I suppose it's just a different way of thinking about the same structure.

**Comment by shawnw** [comment]
Also, semi-related: Running `make indent` on a branch before submitting it as a PR would be nice to enforce.

**Comment by captdeaf** [comment]
I'm down with `make indent`, but man, I want saner rules for it.

```
#define FOO_VAL                                              \
       60 /* This is a really annoying wrap performed by make indent */
```

**Comment by shawnw** [comment]
The old GNU indent did strange stuff sometimes, the current clang-format does strange stuff sometimes. The latter seems to do far fewer strange things though (Plus it's actually maintained).

Not sure what [rules to tweak](https://releases.llvm.org/6.0.0/tools/clang/docs/ClangFormatStyleOptions.html) for that example aside from increasing the column width from 80; nothing seems to directly apply. Maybe `AlignTrailingComments` in a larger context?

**Comment by talvo** [comment]
This current make indent buggers up a lot of carefully hand-aligned stuff, too (like 0x??? `#define`s which are carefully aligned on the x), unfortunately.

**Comment by talvo** [comment]
> Bug fixes and trivial changes? Meh.

One man's "trivial change" is another's "buggering up cross-platform and backwards-compatible behaviour", though, which is why we need this.
